Resources/CoreData/AtomicModules/AtomicEventLoop.js

```diff
@@ -0,0 +1,502 @@
+ * @param {int} delay number of milliseconds to delay
+ * @returns {int} the id of the timer so that it can be stopped in the future.
+ */
+function setInterval(func, delay) {
+    var cb_func;
+    var bind_args;
+    var timer_id;
+    var evloop = EventLoop;
+
+    if (typeof delay !== 'number') {
+        throw new TypeError('delay is not a number');
+    }
+    delay = Math.max(evloop.minimumDelay, delay);
+
+    if (typeof func === 'string') {
+        // Legacy case: callback is a string.
+        cb_func = eval.bind(this, func);
+    } else if (typeof func !== 'function') {
+        throw new TypeError('callback is not a function/string');
+    } else if (arguments.length > 2) {
+        // Special case: callback arguments are provided.
+        bind_args = Array.prototype.slice.call(arguments, 2); // [ arg1, arg2, ... ]
+        bind_args.unshift(this); // [ global(this), arg1, arg2, ... ]
+        cb_func = func.bind.apply(func, bind_args);
+    } else {
+        // Normal case: callback given as a function without arguments.
+        cb_func = func;
+    }
+
+    timer_id = evloop.nextTimerId++;
+
+    evloop.insertTimer({
+        id: timer_id,
+        oneshot: false,
+        cb: cb_func,
+        delay: delay,
+        target: Date.now() + delay
+    });
+
+    return timer_id;
+}
+
+/**
+ * stop a repeated timer that has been set.
+ * @method
+ * @param {int} timer_id the id of the timer that was created via setInterval
+ */
+function clearInterval(timer_id) {
+    var evloop = EventLoop;
+
+    if (typeof timer_id !== 'number') {
+        throw new TypeError('timer ID is not a number');
+    }
+    evloop.removeTimerById(timer_id);
+}
+
+/* custom call */
+/**
+ * custom call to exit out of the current event loop.
+ */
+function requestEventLoopExit() {
+    EventLoop.requestExit();
+}
+
+// Create a closure that will throttle the EventLoop.processTimers function from being executed too often
+function throttleProcessTimers(ms) {
+    var deltaTimer = 0;
+    var pendingOps = false;
+
+    function doThrottle(eventData) {
+        deltaTimer += eventData.timeStep * 1000;
+        if (deltaTimer > ms || pendingOps) {
+            deltaTimer -= ms;
+            // double check to see if we have more timers that expire during this sim
+            // if we do, then we want to process them in the next update, and not 
+            // wait 100ms
+            pendingOps = EventLoop.processTimers();
+        }
+    }
+    return doThrottle;
+}
+
+// Hook into the update event of the engine and process all the timers.  These
+// will be throttled to only run every 100ms
+Atomic.engine.subscribeToEvent('Update', throttleProcessTimers(100));
+
+// Hook into the postUpdate event of the engine and process all the setImmediate calls.  These should
+// really process once the current update loop is completed.
+Atomic.engine.subscribeToEvent('PostUpdate', function () {
+    EventLoop.processImmediates;
```

Comment:
Actually, hold off on accepting this PR.  I did a quick search to see how DukTape handles closures and apparently they are less performant than just creating a prototype object.  I'd like to modify this code to track the state inside a small prototype class instead.  

from a duktape issue:
>Prototype functions are more efficient: closures involve tracking scope objects internally to deal with potentially unknown variable lookups in the inner function (which may, for instance, eval an argument!). Duktape does some primitive detection to avoid tracking scopes for inner functions when possible - e.g. if a function makes direct variable accesses only to its own identifiers, makes no eval() calls, and has no inner functions, its outer scope doesn't need to be tracked. This is not very sophisticated, but covers some basic cases.